Add introduction to Spring Framework (#172)
se-edu/learningresources: contents/java/Spring.md

<frontmatter>
  title: Introduction to Spring Framework
  header: pagetop.md
  footer: footer.md
  head: head.md
  siteNav: mainNav.md
  pageNav: 3
</frontmatter>

<div class="website-content">

{{ booktitle | safe }}

# Introduction to Spring Framework

**Authors: [Liu Yiwen](https://github.com/0blivious)** <br>
Reviewers: []()

<box type="info">
This chapter assumes that the reader has a basic knowledge of Java enterprise.
</box>

<box id="article-toc">

* [What is Spring Framework‎](#what-is-spring-framework)
* [Core Features](#core-features)
  * [The IOC Container](#the-ioc-container)
  * [Spring JDBCTemplate](#spring-jdbctemplate)
  * [Spring AOP](#spring-aop)

> * [Why use Spring‎](#why-use-spring)

Reviewer: The benefits of Spring are not included in the ToC. Although each section is short, maybe it is better to include them for consistency.

* [How to get started‎](#how-to-get-started)
</box>

## What is Spring Framework?

The [official website](https://docs.spring.io/spring/docs/current/spring-framework-reference/overview.html) describes Spring Framework as follows:

>*Spring Framework* is an open-source framework to create Java enterprise applications.
>It provides foundational support for different application architectures, including messaging, transactional data and persistence, and web.

## Core Features

### The IoC Container

> The Spring IoC container is at the core of the Spring Framework. The container will create the objects, wire them together,

Reviewer 2: Would be good to clarify what `IoC` is

Reviewer 2: Could remove the `(DI)` acronym since it isn't used elsewhere in the article

Reviewer 2: Tooltip/popover explaining what dependency injection could be nice https://markbind.org/userGuide/usingComponents.html#popovers

configure them, and manage their complete life cycle from creation till destruction. The Spring container uses dependency injection (DI)
to manage the components that make up an application.
    

> In the Spring framework, the IoC container is represented by the interface ApplicationContext. In order to assemble beans,

Reviewer 2: Is `beans` something in Spring? Might be good to clarify what it is first

Reviewer 2: Could wrap `ApplicationContext` with code

the container uses configuration metadata, which can be in the form of XML configuration or annotations.

Below is one way to manually instantiate a container:
```
    ApplicationContext context = new ClassPathXmlApplicationContext("applicationContext.xml");
```
   
We can then perform dependency injection by using the external XML file.
It makes our code loosely coupled and easier for testing. For example, we can have a simple Employee class

> containing two fields id and name with its setters and getters and a method to display these informations.

Reviewer 3: 'informations'→'information'

```java
    package com.company;  
  
    public class Employee {  
       private int id;  
       private String name;  
       private String city;  
  
       public int getId() {  
           return id;  
       }  
       public void setId(int id) {  
           this.id = id;  
       }  
       public String getName() {  
           return name;  
       }  
       public void setName(String name) {  
           this.name = name;  
       }  
       void display(){  
           System.out.println(id+" "+name);  
       }
    }
```
    
And the corresponding `applicationContext.xml` file could be:

> ```

Reviewer: Can add syntax highlighting for xml files?
```suggestion
```xml
```

    <?xml version="1.0" encoding="UTF-8"?>  
    <beans  
    xmlns="http://www.springframework.org/schema/beans"  
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"  
    xmlns:p="http://www.springframework.org/schema/p"  
    xsi:schemaLocation="http://www.springframework.org/schema/beans  
                http://www.springframework.org/schema/beans/spring-beans-3.0.xsd">  
  

>     <bean id="obj" class="com.company.Employee">  

Reviewer: Would it be better to add some indentation here so that it is easier for the reader to understand? Also maybe can omit the xml namespaces to only highlight the important part?

    <property name="id">  
    <value>20</value>  
    </property>  
    <property name="name">  
    <value>Amy</value>  
    </property>  
    </bean>  
    </beans>
```

> We are providing the information into the bean by this file. The property element invokes the setter method. The value subelement of property will assign the specified value.

Reviewer 3: This paragraph can benefit from clearly defining the actors at play.
Consider something like:
> We can populate the `Employee` bean with the data provided in `applicationContext.xml`. 

It might not be necessary to explain the bits about invoking the setters.

   
To better understand how this dependency injection work, we can have a `Test.java` looks like the following:
```java
    public class Test {  
        public static void main(String[] args) {   

>             Resource r=new ClassPathResource("applicationContext.xml");  

Reviewer 3: Consider using the same coding style throughout the article. 
```
Resource r = new ClassPathResource("applicationContext.xml");
        ^ ^
```

            BeanFactory factory=new XmlBeanFactory(r);  
          
            Employee e=(Employee)factory.getBean("obj");  
            s.display();
        }  
    }  
```
The out put of the above code will be `20 Amy`.

### Spring JdbcTemplate

> Spring JdbcTemplate is a powerful mechanism to connect to the database and execute SQL queries. It internally uses

Reviewer 2: explanation of jdbc would be nice

> JDBC API, but eliminates a lot of problems of it.

Reviewer 3: Grammar and tone nit.

'eliminates ... of it' → 'alleviates ... associated with it'

> The problems of traditional JDBC API are listed as follows:

Reviewer 3: 'The' → 'Some'

- We need to write a lot of code before and after executing the query, such as creating connection,
statement, closing resultset, connection etc.
- We need to perform exception handling code on the database logic.
- We need to handle transaction.
- Repetition of all these codes from one to another database logic is a time consuming task.

> Spring JdbcTemplate eliminates all the above mentioned problems of JDBC API.

Reviewer 3: Duplication of line 125.

It provides you methods to write the queries directly, so it saves a lot of work and time.

> ![Spring JdbcTemplate](springJDBC.png "How much Spring saves you")

Reviewer: Maybe can add a label to the image, or write the table in markdown to make the style consistent. Currently the image appears a little bit too large.

Reviewer 3: Consider rephrasing 'SQL' to 'Writing SQL queries'

### Spring AOP

> Aspect Oriented Programming (AOP) compliments OOPs in the sense that it provides modularity in aspect instead of class.

Reviewer 2: expand `OOP` first

Reviewer 2: Might restructure it such that you start with the examples to ease the reader into the concept of `AOP`

> Aspect Oriented Programming (AOP) compliments OOPs in the sense that it provides modularity in aspect instead of class.

Reviewer 3: Consider rephrasing by introducing the cause before the effect for clarity.
Something like: 
`AOP is ... and has the effect of ... because ...`

AOP breaks the program logic into distinct parts called concerns. A cross-cutting concern is a concern that can affect
the whole application and should be centralized in one location in code as much as possible. Examples of these concerns
include transaction management, authentication, logging, security etc.

> By separating application business logic from system services, Spring Framework supports Aspect Oriented Programming

Reviewer 2: > It provides the way to dynamically add the
cross-cutting concern before, after or around the actual logic using simple
pluggable configurations.

Does pluggable configurations here refer to the XML files?

> It also allows easier to maintainance of the code.

How so?

and enables cohesive development. It provides the way to dynamically add the
cross-cutting concern before, after or around the actual logic using simple
pluggable configurations. It also allows easier to maintainance of the code. You can add/remove concerns
without recompiling complete sourcecode simply by changing configuration file by applying aspects using XML configuration.

## Why use Spring?

Spring makes programming Java quicker, easier, and safer for everybody.

> Like all other general Java frameworks, Spring helps us focus on the core task rather than the boilerplate associated with it.

Reviewer 3: 'Like all other general Java frameworks' → 'Similar to other Java frameworks (e.g. Grails, Play)'

Apart from that, Spring has other advantages like:

### Benefit 1: Modularity

> Spring provides different modules to achieve different services and functionality for development of application.

Reviewer 2: Some examples of these modules would be nice, or a brief back-reference if they were mentioned previously

Reviewer 2: `for application development`

These modules are designed in such a way that no module is dependent on the others, except Spring core module.

> Thus, we can optionally include one or more Spring projects depending on the need.

Reviewer 3: Consider explaining the benefits of modularity, e.g. smaller footprint, etc.

> ### Benefit 2: Versatile

Reviewer 2: Perhaps `easily integrated into existing projects`? `versatile` seems a little too general here

> Spring is designed to be used with all other frameworks of Java, you can use ORM, Struts, Hibernate and other frameworks

Reviewer 3: Consider explaining ORM. Also how the other frameworks can help you.

of Java together with Spring. Spring framework do not impose any restriction on the frameworks to be used together.

### Benefit 3: Strong Community Support
Spring is an open source framework led by Pivotal Software and backed by a large consortium of organizations and individual

> developers. This has meant that it remains contextual and often futuristic, as evident by the number of projects under its umbrella.

Reviewer 2: a chart or some numbers here could be nice to back this up

> developers. This has meant that it remains contextual and often futuristic, as evident by the number of projects under its umbrella.

Reviewer 3: 'remains contextual and futuristic' → 'remains relevant'

### Benefit 4: Usability
One of the key aspects of any framework's popularity is how easy it is for developers to use it.

> Spring through multiple configuration options and Convention over Configuration makes it really easy for developers to

Reviewer 2: `Convention over Configuration` seems contrary to `multiple configuration options`.
Did you mean that it is simple to set up, but powerful to extend?

start and then configure exactly what they need.

> Projects like Spring Boot have made bootstrapping a complex Spring project almost trivial. Not to mention, it has excellent documentation and tutorials to help anyone get on-boarded.

Reviewer 3: Link to Spring Boot?

<frontmatter>
  title: Introduction to Spring Framework
  header: pagetop.md
  footer: footer.md
  head: head.md
  siteNav: mainNav.md
  pageNav: 3
</frontmatter>

<div class="website-content">

{{ booktitle | safe }}

# Introduction to Spring Framework

**Authors: [Liu Yiwen](https://github.com/0blivious)** <br>

> Reviewers: [ANG ZE YU](https://github.com/ang-zeyu), [LUM KA FAI JEFFRY](https://github.com/j-lum), [NI TIANZHEN](https://github.com/niqiukun)

Reviewer 4: Should the names be in all caps?

which can be in the form of XML configuration or annotations.

Below is one way to manually instantiate a container:
```java
ApplicationContext context = new ClassPathXmlApplicationContext("applicationContext.xml");
```

For example, say we have an Employee class like so:
```java
package com.company;  
  
public class Employee {  
    private int id;  
    private String name;  
    private String city;  
  
    public int getId() {  
        return id;  
    }  
    public void setId(int id) {  
        this.id = id;  
    }  
    public String getName() {  
        return name;  
    }  
    public void setName(String name) {  
        this.name = name;  
    }  
    void display(){  
        System.out.println(id+" "+name);  
    }
}        
```

We can then populate the Employee bean with the data provided in the following `applicationContext.xml` file:
This makes our code loosely coupled and easier for testing. 
```xml
<bean id="employeeAmy" class="com.company.Employee">  
    <property name="id">  
        <value>20</value>  
    </property>  
    <property name="name">  
        <value>Amy</value>  
    </property>  
</bean>  
```
   

> To better understand how this dependency injection works, we can have a `Test.java` looks like the following:

Reviewer 4: Unexplained term: dependency injection

> To better understand how this dependency injection works, we can have a `Test.java` looks like the following:

Reviewer 4: Also consider this flow:
First, give the employee code.
Then, give the xml file that explains how to populate employee objects.
Then, show how the xml file is used to instantiate the container, which will result in objects being populated.
After that, show the test code, which proves the objects have been populated.

```java
Resource r = new ClassPathResource("applicationContext.xml");  
BeanFactory factory = new XmlBeanFactory(r);  
          
Employee e = (Employee)factory.getBean("obj");  
e.display();
```

The output of the above code would then be `20 Amy`.

### Spring JdbcTemplate

> The Spring JdbcTemplate is a powerful mechanism to connect to the database and execute SQL queries. Traditionally, we

Reviewer 4: `from Oracle Corporation` is unnecessary?

need to use Java Database Connectivity (JDBC) API to access tabular data stored in any relational database.
It is a part of JavaSE (Java Standard Edition), from Oracle Corporation.
    
In a traditional JDBC API:
- a typical SQL query would involve quite some amount of boilerplate code, such as creating connection,
statement, closing resultset, connection etc.
- exception handling has to be manually written out.
- transactions need to be manually created.

The Spring JdbcTemplate provides methods to write the queries directly, hence, saving time in writing such boilerplate code.

Task | Spring | You
-----|--------|----
Connection Management | :far-check-circle: | 
Writing SQL Queries | | :far-check-circle: 
Statement Management | :far-check-circle: |
ResultSet Management | :far-check-circle: |
Row Data Retrieval | | :far-check-circle: 
Parameter Declaration | | :far-check-circle: 
Parameter Setting | :far-check-circle: |
Transaction Management | :far-check-circle: |
Exception Handling | :far-check-circle: |

> 

Reviewer 4: It's nice if you can give example code for both, for a comparison. OK if you can't find such code easily.

### Spring AOP

Spring Framework addresses
<tooltip content="Concerns that can affect the whole application and should be centralized in one location in code as much as possible. Examples of these concerns include transaction management, authentication, logging, security etc.">

> <i>cross-cutting concerns</i></tooltip> by supporting Aspect Oriented Programming (AOP).

Reviewer 4: Link `AOP` to a good explanation of it e.g., wikipedia page

We can then write aspect class annotated with `@Aspect` annotation and write point-cut expressions to log program information from
<tooltip content="Join point is a point of execution of the program, such as the execution of a method or the handling of an exception.">
<i>joint-point</i>
</tooltip> methods.
```java
@Aspect
public class EmployeeCRUDAspect {
      
    @Before("execution(* EmployeeManager.getEmployeeById(..))") // point-cut expressions
    public void logBefore(JoinPoint joinPoint) {
        System.out.println("EmployeeCRUDAspect.logBefore() : "
                + joinPoint.getSignature().getName());
    }
    
    @After("execution(* EmployeeManager.getEmployeeById(..))") // point-cut expressions
    public void logAfter(JoinPoint joinPoint) {
        System.out.println("EmployeeCRUDAspect.logAfter() : "
                + joinPoint.getSignature().getName());
    }
}
```
    
And the corresponding `EmployeeManager.java` file could be:
```java
@Component
public class EmployeeManager
{
    public EmployeeDTO getEmployeeById(Integer employeeId) {
        System.out.println("Method getEmployeeById() called");
        return new EmployeeDTO();
    }
}

> ```

Reviewer 4: Use Egyptian style braces?
Also, if you took these code examples from somewhere, cite the source.

**Authors: [Liu Yiwen](https://github.com/0blivious)** <br>
Reviewers: []()

<box type="info">
This chapter assumes that the reader has a basic knowledge of Java enterprise.
</box>

<box id="article-toc">

* [What is Spring Framework‎](#what-is-spring-framework)
* [Core Features](#core-features)
  * [The IOC Container](#the-ioc-container)
  * [Spring JdbcTemplate](#spring-jdbctemplate)
  * [Spring AOP](#spring-aop)
* [Why use Spring‎](#why-use-spring)
  * [Benefit 1: Modularity](#benefit-1-modularity)
  * [Benefit 2: Easy Integration](#benefit-2-easy-integration)
  * [Benefit 3: Strong Community Support](#benefit-3-strong-community-support)
  * [Benefit 4: Usability](#benefit-4-usability)
* [How to get started‎](#how-to-get-started)
</box>

## What is Spring Framework?

The [official website](https://docs.spring.io/spring/docs/current/spring-framework-reference/overview.html) describes Spring Framework as follows:

>*Spring Framework* is an open-source framework to create Java enterprise applications.
>It provides foundational support for different application architectures, including messaging, transactional data and persistence, and web.

## Core Features

> ### The IoC Container

Reviewer 2: feels a little heavy

Inversion of Control (IoC) is a design principle. As the name suggests, it is used to invert

> different kinds of controls in object-oriented design to achieve loose coupling.

Reviewer 2: `loose` -> `looser`

Here, controls refer to any additional responsibilities a class has,

> other than its main responsibility. This include control over the flow of an application,

Reviewer 2: ~This include~ -> `For example, this may include the control ... application and object creation` ~dependent object creation and binding~ ( let's stick to simple examples )

and control over the flow of an object creation or dependent object creation and binding.

> The Spring IoC container is at the core of the Spring Framework. The container will create the objects, wire them together,

Reviewer 2: The container will create the objects - does `objects` here refer to beans?
Could first introduce the concept of beans if so

<tooltip content="These objects are are called beans in Spring Framework. A bean is an object that is instantiated,
assembled, and otherwise managed by a Spring IoC container.">
<i>objects</i>
</tooltip> 
, wire them together,
configure them, and manage their complete life cycle from creation till destruction. The Spring container uses
<tooltip content="Dependency injection is a way of providing a class with the required services.">
<i>dependency injection</i>

> </tooltip> to manage the components that make up an application.

Reviewer 4: Not sure if the mentioning of _dependency injection_ has any value i.e., minimize technical terms.

> </tooltip> to manage the components that make up an application.

Reviewer 4: A lighter way to explain this:

>Normally, objects in a Java application often have to take on  additional responsibilities such as creating other objects , wiring them together, configuring them, and managing their life cycle. When using the Spring framework for an application, the framework takes care of many of those responsibilities. This design approach is called _inversion of control (IoC)_ and Spring achieves it using something called the _IoC container_.

IoC is common trait of frameworks by the way (as you learned in CS2103); not specific to Spring.

    
In the Spring framework, the IoC container is represented by the interface `ApplicationContext`.
The beans are created with the configuration metadata that you supply to the container,

> which can be in the form of XML configuration or annotations.

Reviewer 4: Unexplained term: bean

ApplicationContext context = new ClassPathXmlApplicationContext("applicationContext.xml");
```
   

> We can then perform dependency injection by using the external XML file.

Reviewer 2: Repeat of the below `We can then populate the Employee bean with the data provided in the following applicationContext.xml file:`

> It makes our code loosely coupled and easier for testing. For example, we can have a simple Employee class

Reviewer 2: For example, say we have an `Employee` class like so: ~two fields id and name with its setters and getters and a method to display these information.~

For example, say we have an Employee class like so:
```java
package com.company;  
  
public class Employee {  
    private int id;  
    private String name;  
    private String city;  
  
    public int getId() {  
        return id;  
    }  
    public void setId(int id) {  
        this.id = id;  
    }  
    public String getName() {  
        return name;  
    }  
    public void setName(String name) {  
        this.name = name;  
    }  
    void display(){  
        System.out.println(id+" "+name);  
    }
}        
```

We can then populate the Employee bean with the data provided in the following `applicationContext.xml` file:
This makes our code loosely coupled and easier for testing. 
```xml
<bean id="obj" class="com.company.Employee">  
    <property name="id">  
        <value>20</value>  
    </property>  
    <property name="name">  
        <value>Amy</value>  
    </property>  
</bean>  

> ```

Reviewer 4: `obj` -> use a better name?

> To better understand how this dependency injection work, we can have a `Test.java` looks like the following:

Reviewer 2: Could show just the code excerpt

```java
Resource r = new ClassPathResource("applicationContext.xml");  
BeanFactory factory = new XmlBeanFactory(r);  
          
Employee e = (Employee)factory.getBean("obj");  
s.display();
```

since the `Test` class only serves to contain `main()`

Reviewer 2: `works`

public class Test {  
    public static void main(String[] args) {   
        Resource r = new ClassPathResource("applicationContext.xml");  
        BeanFactory factory = new XmlBeanFactory(r);  
          
        Employee e = (Employee)factory.getBean("obj");  

>         s.display();

Reviewer 2: `e.display()`

    }  
}  
```
The out put of the above code will be `20 Amy`.

### Spring JdbcTemplate

> Spring JdbcTemplate is a powerful mechanism to connect to the database and execute SQL queries. Traditionally, we

Reviewer 2: `The Spring JdbcTemplate`
~It is a part of JavaSE (Java Standard Edition), from Oracle Corporation.~ ( since the reader assumption is of java enterprise )

need to use Java Database Connectivity (JDBC) API to access tabular data stored in any relational database.
It is a part of JavaSE (Java Standard Edition), from Oracle Corporation.
    

> Some problems of traditional JDBC API are listed as follows:

Reviewer 2: Suggestion on the structure:
`In a traditional JDBC API,`

> - We need to write a lot of code before and after executing the SQL query, such as creating connection,

Reviewer 2: `a typical SQL query would involve quite some amount of boilerplate code, such as ...`
`exception handling has to be written`
`transactions need to be manually created`

statement, closing resultset, connection etc.
- exception handling has to be manually written out.
- transactions need to be manually created.

The Spring JdbcTemplate provides methods to write the queries directly, hence, saving time in writing such boilerplate code.

Task | Spring | You
-----|--------|----

> Connection Management | :heavy_check_mark: | 

Reviewer 4: fontawesome icons give a more professional look than emoji. e.g., `:far-check-circle:`

https://markbind.org/userGuide/formattingContents.html#using-font-awesome-icons

> - Repetition of all these codes from one to another database logic is a time consuming task.

Reviewer 2: repeat of `saving a lot of work and time`

> It provides methods to write the queries directly, saving a lot of work and time.

Reviewer 2: `The Spring JdbcTemplate provides ... hence, saving time in writing such boilerplate code`

-----|--------|----
Connection Management | :heavy_check_mark: | 
Writing SQL Queries | | :heavy_check_mark: 
Statement Management | :heavy_check_mark: |
ResultSet Management | :heavy_check_mark: |
Row Data Retrieval | | :heavy_check_mark: 
Parameter Declaration | | :heavy_check_mark: 
Parameter Setting | :heavy_check_mark: |
Transaction Management | :heavy_check_mark: |
Exception Handling | :heavy_check_mark: |

### Spring AOP

Aspect Oriented Programming (AOP) is a programming paradigm that aims to increase modularity by allowing the separation of

> <tooltip content="A cross-cutting concern is a concern that can affect the whole application and should be centralized in one location in code as much as possible.">

Reviewer 2: ~A cross-cutting concern is a~ `Concerns that ...` 
Could move the examples in here as well ( could use a [popover](https://markbind.org/userGuide/usingComponents.html#popovers) for more space )

without modifying the code itself, instead separately specifying which code is modified via a "pointcut"
specification. This allows behaviors that are not central to the business logic (such as logging) to be added to a program

> without cluttering the core code. AOP forms a basis for aspect-oriented software development.

Reviewer 4: >AOP forms a basis for aspect-oriented software development.

This doesn't make sense. AOP is simply the abbreviation for _Aspect-Oriented Programming_ (which is same as aspect-oriented software development).  AOP is technique invented elsewhere and being used by Spring (just like IoC)

> Spring Framework supports AOP and enables cohesive development, by providing ways to dynamically add the

Reviewer 4: >Spring Framework supports AOP and enables cohesive development

Hmm... what's _cohesive development_?

cross-cutting concern before, after or around the actual logic using simple
pluggable configurations. 

For example, say we wish to log something into the console everytime before we call the method `getEmployeeById`.

> We can then write aspect class annotated with `@Aspect` annotation and write point-cut expressions to log program information from joint points

Reviewer 2: extra `joint points`

> By separating application business logic from system services, Spring Framework supports AOP

Reviewer 2: ~By separating application business logic from system services~ ( repeat of above )
Rather,

`Spring Framework supports AOP and ... by providing ways to dynamically...`

> For example, we wish to log something into the console everytime before we call the method `getEmployeeById`.

Reviewer 2: `For example, say we wish ... `

We can then write aspect class annotated with `@Aspect` annotation and write point-cut expressions to match
<tooltip content="Join point is a point of execution of the program, such as the execution of a method or the handling of an exception.">

> <i>joint-point</i>

Reviewer 2: ~match~ `to log program information from joint points`

</tooltip> methods.
```java
@Aspect
public class EmployeeCRUDAspect {
      
    @Before("execution(* EmployeeManager.getEmployeeById(..))") // point-cut expressions
    public void logBefore(JoinPoint joinPoint) {
        System.out.println("EmployeeCRUDAspect.logBefore() : "
                + joinPoint.getSignature().getName());
    }
    
    @After("execution(* EmployeeManager.getEmployeeById(..))") // point-cut expressions
    public void logAfter(JoinPoint joinPoint) {
        System.out.println("EmployeeCRUDAspect.logAfter() : "
                + joinPoint.getSignature().getName());
    }
}
```
    
And the corresponding `EmployeeManager.java` file could be:
```java
@Component
public class EmployeeManager
{
    public EmployeeDTO getEmployeeById(Integer employeeId) {
        System.out.println("Method getEmployeeById() called");
        return new EmployeeDTO();
    }
}
```

> In above example, `logBefore()` will be executed before `getEmployeeById()` method because it matches the join-point expression.

Reviewer 2: `In the above`
`before the getEmployeeById() method`
`matches the join point specified earlier`

Similarly, `logAfter()` will be executed after `getEmployeeById()`.

> The out put of the above code will be:

Reviewer 2: `output of the above code would then be:`

Reviewer 2: `output`

```
EmployeeCRUDAspect.logBefore() : getEmployeeById
Method getEmployeeById() called
EmployeeCRUDAspect.logAfter() : getEmployeeById
```

> The declaration of aspects and advices can also go into XML files so we do not need to write the annotation in our

Reviewer 2: Is `advices` something similar to aspects? Could add a tooltip if so

`could also be written directly in the XML templates` ~we do not need to write the annotation in our..~

> source code. This allows easier maintainance of the code as we can add/remove concerns 

Reviewer 2: `allows for`

> by changing XML configuration file without recompiling source code. The above configuration can be written in XML file as:

Reviewer 2: would it not be easier to read (and hence maintain) as well?

Paragraph break +  `For example, the above ...` ~can~ `could be written in a XML ...`

For example, the above configuration could be written in a XML file as:
```xml
 <aop:config>
    <!-- Spring AOP Pointcut definitions -->
    <aop:pointcut id="CRUDOperation" expression="execution(* EmployeeManager.*(..))" />
 
    <!-- Spring AOP aspect 1 -->
    <aop:aspect id="EmployeeCRUD" ref="crudAspectBean">
             
        <!-- Spring AOP advises -->
        <aop:before pointcut-ref="CRUDOperation" method="logBefore" />
        <aop:after pointcut-ref="CRUDOperation" method="logAfter" />
             
    </aop:aspect>

    <!-- Spring AOP aspect instances -->
    <bean id="crudAspectBean" class="com.aop.EmployeeCRUDAspect" />
     
    <!-- Target Object -->
    <bean id="employeeManager" class="com.aop.EmployeeManagerImpl" />
</aop:config>

> ```

Reviewer 4: We should not write as if AOP is a Spring thing. Rather, Spring is _using_ AOP.
Given that, try not to go too much into AOP. If anything, it should be separate chapter by itself.

## Why Use Spring?

> Spring makes programming Java quicker, easier, and safer for everybody.

Reviewer 4: Well, not for _everybody_ :-) You may want to comment on what kind of applications Spring targets (Web-based enterprise applications?)

Similar to other general Java frameworks (e.g. [Grails](https://grails.org/), [Play](https://www.playframework.com/)), Spring helps us focus on the core task rather than the boilerplate associated with it.

> Apart from that, Spring has other advantages like:

Reviewer 4: >Spring has other advantages like:

Best not to set up an open-ended sentence that goes on for many sections. Something like this?

>Spring has several other advantages, described in the sections below.

> ### Benefit 1: Modularity

Reviewer 4: Earlier you said _advantages_ but here you use _benefit_. Use the same term.

> Similar to other general Java frameworks (e.g. Grails, Play), Spring helps us focus on the core task rather than the boilerplate associated with it.

Reviewer 2: link to grails, play would be nice

> These modules are designed in such a way that no module is dependent on the others, except Spring core module.

Reviewer 2: `except the`

Spring provides different modules to achieve different services and functionality for development of application.
These modules are designed in such a way that no module is dependent on the others, except the Spring core module.
Thus, we can optionally include one or more Spring projects depending on the need. This makes Spring lightweight
and require less configuration.

![Spring Framework Overview](spring-overview.png "Overview of the Spring Framework")

_Figure 1. Overview of the Spring Framework_

### Benefit 2: Easy Integration

> Spring is designed to be used compatible with many other frameworks of Java, for example

Reviewer 2: `to be compatible`

> and require less configuration.

Reviewer 2: Hmm not too familiar with Spring modules - how does modularity end up in requiring less configuration?

Author: Sth like if you are forced to use the whole framework, you will need to complete all the configuration even for the part that you will not use.

> Spring is designed to be used with all other frameworks of Java, you can use

Reviewer 2: `to be` ~used~ `compatible`
`all` seems overbearing -> `many`
`frameworks of Java, for example Struts and Hibernate.` ~and other frameworks of Java together with Spring~

<tooltip content="Struts is an open source framework that extends the Java Servlet API and employs a Model, View, Controller (MVC) architecture.">
<i>Struts </i>
</tooltip>and
<tooltip content="Hibernate is an object-relational mapping tool for the Java programming language. It provides a framework for mapping an object-oriented domain model to a relational database.">
<i>Hibernate</i>

> </tooltip>. Spring framework do not impose any restriction on the frameworks to be used together.

Reviewer 2: `It does not`

### Benefit 3: Strong Community Support
Spring is an open source framework led by [Pivotal](https://tanzu.vmware.com/) Software and backed by a large consortium of organizations and individual
developers. This means that it remains relevant,

> as evident by the ever increasing number of projects under its umbrella, which is ever increasing. 

Reviewer 2: ~which is ever increasing~

> Spring is an open source framework led by Pivotal Software and backed by a large consortium of organizations and individual

Reviewer 2: link to Pivotal Software would be nice

> developers. This has meant that it remains relevant,

Reviewer 2: `This means that`
`as evident by the ever increasing number of ...`

<frontmatter>
  title: Introduction to Spring Framework
  header: pagetop.md
  footer: footer.md
  head: head.md
  siteNav: mainNav.md
  pageNav: 3
</frontmatter>

<div class="website-content">

{{ booktitle | safe }}

# Introduction to Spring Framework

**Authors: [Liu Yiwen](https://github.com/0blivious)** <br>
Reviewers: []()

<box type="info">
This chapter assumes that the reader has a basic knowledge of Java enterprise.
</box>

<box id="article-toc">

* [What is Spring Framework‎](#what-is-spring-framework)
* [Core Features](#core-features)
  * [The IOC Container](#the-ioc-container)

>   * [Spring JDBCTemplate](#spring-jdbctemplate)

Reviewer: Can standardise the capitalisation, i.e. `JdbcTemplate`

These beans are created with the configuration metadata that you supply to the container,
which can be in the form of XML configuration or annotations.

Below is one way to manually instantiate a container:
```java
ApplicationContext context = new ClassPathXmlApplicationContext("applicationContext.xml");
```
   
We can then perform dependency injection by using the external XML file.
It makes our code loosely coupled and easier for testing. For example, we can have a simple Employee class
containing two fields id and name with its setters and getters and a method to display these information.
```java
package com.company;  
  
public class Employee {  
    private int id;  
    private String name;  
    private String city;  
  
    public int getId() {  
        return id;  
    }  
    public void setId(int id) {  
        this.id = id;  
    }  
    public String getName() {  
        return name;  
    }  
    public void setName(String name) {  
        this.name = name;  
    }  
    void display(){  
        System.out.println(id+" "+name);  
    }
}        
```
We can then populate the Employee bean with the data provided in the following `applicationContext.xml` file:

> ```xml

Reviewer: Possible to add intentation for this file? It would be easier to read.

Writing SQL Queries | | :heavy_check_mark: 
Statement Management | :heavy_check_mark: |
ResultSet Management | :heavy_check_mark: |
Row Data Retrieval | | :heavy_check_mark: 
Parameter Declaration | | :heavy_check_mark: 
Parameter Setting | :heavy_check_mark: |
Transaction Management | :heavy_check_mark: |
Exception Handling | :heavy_check_mark: |

### Spring AOP

Aspect Oriented Programming (AOP) is a programming paradigm that aims to increase modularity by allowing the separation of
<tooltip content="A cross-cutting concern is a concern that can affect the whole application and should be centralized in one location in code as much as possible.">
<i>cross-cutting concerns</i></tooltip>. Examples of these concerns include transaction management, authentication, logging, security etc. 
It does so by adding additional behavior to existing code
without modifying the code itself, instead separately specifying which code is modified via a "pointcut"
specification. This allows behaviors that are not central to the business logic (such as logging) to be added to a program
without cluttering the core code. AOP forms a basis for aspect-oriented software development.

> By separating application business logic from system services, Spring Framework supports Aspect Oriented Programming

Reviewer: ```suggestion
By separating application business logic from system services, Spring Framework supports AOP
```

and enables cohesive development. It provides the way to dynamically add the
cross-cutting concern before, after or around the actual logic using simple
pluggable configurations. 

For example, we wish to log something into the console everytime before we call the method `getEmployeeById`.

> We can then write aspect class annotated with @Aspect annotation and write point-cut expressions to match

Reviewer: ```suggestion
We can then write aspect class annotated with `@Aspect` annotation and write point-cut expressions to match
```

<tooltip content="Joinpoint is a point of execution of the program, such as the execution of a method or the handling of an exception..">

> <i>joint-point</i>

Reviewer: Typo?

    public void logAfter(JoinPoint joinPoint) {
        System.out.println("EmployeeCRUDAspect.logAfter() : "
                + joinPoint.getSignature().getName());
    }
}
```
    
And the corresponding `EmployeeManager.java` file could be:
```java
@Component
public class EmployeeManager
{
    public EmployeeDTO getEmployeeById(Integer employeeId) {
        System.out.println("Method getEmployeeById() called");
        return new EmployeeDTO();
    }
}
```
In above example, `logBefore()` will be executed before `getEmployeeById()` method because it matches the join-point expression.
Similarly, `logAfter()` will be executed after `getEmployeeById()`.
The out put of the above code will be:
```
EmployeeCRUDAspect.logBefore() : getEmployeeById
Method getEmployeeById() called
EmployeeCRUDAspect.logAfter() : getEmployeeById
```

The declaration of aspects and advices can also go into XML files so we do not need to write the annotation in our
source code. This allows easier maintainance of the code as we can add/remove concerns 
by changing XML configuration file without recompiling source code. The above configuration can be written in XML file as:

> ```xml

Reviewer: Seems that the indentation in this file is inconsistent. Let's keep it to 4 spaces.

 <aop:config>
        <!-- Spring AOP Pointcut definitions -->
        <aop:pointcut id="CRUDOperation"
            expression="execution(* EmployeeManager.*(..))" />
 
        <!-- Spring AOP aspect 1 -->
        <aop:aspect id="EmployeeCRUD" ref="crudAspectBean">
             
            <!-- Spring AOP advises -->
            <aop:before pointcut-ref="CRUDOperation" method="logBefore" />
            <aop:after pointcut-ref="CRUDOperation" method="logAfter" />
             
        </aop:aspect>

        <!-- Spring AOP aspect instances -->
        <bean id="crudAspectBean" class="com.aop.EmployeeCRUDAspect" />
     
        <!-- Target Object -->
        <bean id="employeeManager" class="com.aop.EmployeeManagerImpl" />
</aop:config>
```

> ## Why use Spring?

Reviewer 5: Why Use Spring?

<i>Struts</i>
</tooltip>, 
<tooltip content="Hibernate is an object-relational mapping tool for the Java programming language. It provides a framework for mapping an object-oriented domain model to a relational database.">
<i>Hibernate</i>
</tooltip>, and other frameworks
of Java together with Spring. Spring framework do not impose any restriction on the frameworks to be used together.

### Benefit 3: Strong Community Support
Spring is an open source framework led by Pivotal Software and backed by a large consortium of organizations and individual
developers. This has meant that it remains relevant,
as evident by the number of projects under its umbrella, which is ever increasing. 

### Benefit 4: Usability
One of the key aspects of any framework's popularity is how easy it is for developers to use it.
Projects like [Spring Boot](https://spring.io/projects/spring-boot) have made bootstrapping a complex Spring project almost trivial.
Spring thus is very easy to start but also powerful to extend to what programmers exactly need by providing multiple configuration options.
Not to mention, it has excellent documentation and tutorials to help anyone get on-boarded.

> ## How to get started‎

Reviewer 5: Getting Started
